### PR TITLE
Rewrite `update-since-todo.sh` in Python

### PR DESCRIPTION
**Pull request by basil**
Daniel's feedback in #6398 (to split the existing output into multiple sections with a separate section for `@RestrictedSince`) is painful to implement in a shell script. By rewriting this as a Python script, it becomes easier for new contributors to improve this script, especially the younger generation of programmers for whom Python is their first scripting language.

### Proposed changelog entries

N/A

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [x] (If applicable) Jira issue is well described
- [x] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change) and are in the imperative mood. [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [x] Appropriate autotests or explanation to why this change has no tests
- [x] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadoc, as appropriate. 
- [x] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@mention

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [x] There are at least 2 approvals for the pull request and no outstanding requests for change
- [x] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [x] Changelog entries in the PR title and/or `Proposed changelog entries` are accurate, human-readable, and in the imperative mood
- [x] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

### CODE REVIEW

**Comment by daniel-beck** [comment]
> new contributors to improve this script, especially the younger generation of programmers for whom Python

For a Jenkins project thing, wouldn't Groovy be the scripting language of choice? Between Pipeline, Job DSL, the script console, advanced Jenkins users probably have a fairly good baseline familiarity with it.

I realize the work is done and it's not worth switching again, but just a thought for future endeavors of this kind.

**Comment by basil** [comment]
I wanted this to be something anyone on any reasonable Unix-like system can run without any installation or setup. Groovy doesn't meet that criterion.

**Review by timja — APPROVED** [review]

**Review by NotMyFault — APPROVED** [review]
Thanks for your alternative approach, basil!

**Comment by basil** [comment]
This PR is now ready for merge. We will merge it after approximately 24 hours if there is no negative feedback. Please see the [merge process documentation](https://github.com/jenkinsci/jenkins/blob/master/docs/MAINTAINERS.adoc#merge-process) for more information about the merge process. Thanks!

**Comment by basil** [comment]
> Thanks for your alternative approach, basil!

You're welcome, but this isn't an alternative approach to your PR, as it doesn't add support for `@RestrictedSince` (nor do I intend to do so). It's a rewrite of the original script that should make addressing Daniel's feedback in your PR more feasible.

File: update-since-todo.py
Comment (basil) [reply update-since-todo.py:27]: Yeah, that is my understanding as well.

**Pull request merged**
